### PR TITLE
Added TryWrite and TryRead

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
These two APIs will allow us to optimize some code currently using Read/Write

### CODE REVIEW

File: src/System.Slices/System/SpanExtensions_binary.cs [hunk 1]
```diff
@@ -62,6 +62,36 @@ public static partial class SpanExtensionsLabs
...
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TryRead<[Primitive]T>(this ReadOnlySpan<byte> slice, out T value)
+            where T : struct
+        {
+            if (Unsafe.SizeOf<T>() > (uint)slice.Length) {
+                value = default(T);
+                return false;
+            }
+            value = slice.Cast<byte, T>()[0];
```
Comment (jkotas) [inline src/System.Slices/System/SpanExtensions_binary.cs]: More efficient way to do this is: `value = Unsafe.As<byte, T>(ref slice.DangerousGetPinnableReference())`

Comment (jkotas) [inline src/System.Slices/System/SpanExtensions_binary.cs]: The usual disclaimer: This can crash with alignment fault on ARM if T==float or double, etc.

Comment (KrzysztofCwalina) [inline src/System.Slices/System/SpanExtensions_binary.cs]: thanks. will change.

File: src/System.Slices/System/SpanExtensions_binary.cs [hunk 2]
```diff
@@ -58,7 +58,37 @@ public static partial class SpanExtensionsLabs
...
             Contract.RequiresInInclusiveRange(Unsafe.SizeOf<T>(), (uint)slice.Length);
-            return slice.Cast<byte, T>()[0];
+            return Unsafe.As<byte, T>(ref slice.DangerousGetPinnableReference());
+        }
+
+        /// <summary>
+        /// Reads a structure of type T out of a slice of bytes.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TryRead<[Primitive]T>(this ReadOnlySpan<byte> slice, out T value)
```
Comment (ahsonkhan) [inline src/System.Slices/System/SpanExtensions_binary.cs]: What does [Primitive] do? Does it enforce that the generic T can only be primitive types?

File: src/System.Slices/System/SpanExtensions_binary.cs [hunk 3]
```diff
@@ -72,6 +102,21 @@ public static partial class SpanExtensionsLabs
...
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TryWrite<[Primitive]T>(this Span<byte> slice, T value)
+            where T : struct
+        {
+            if (Unsafe.SizeOf<T>() > (uint)slice.Length) {
+                return false;
+            }
+            Span<T> castedSlice = slice.Cast<byte, T>();
+            castedSlice[0] = value;
```
Comment (ahsonkhan) [inline src/System.Slices/System/SpanExtensions_binary.cs]: Always write to index 0? Is it useful to have an overload which writes to index n?

File: tests/System.Slices.Tests/CastTests.cs [hunk 4]
```diff
@@ -9,6 +9,48 @@ namespace System.Slices.Tests
...
+        public void TryReadReadOnlySpan()
+        {
+            uint value;
+            ReadOnlySpan<byte> buffer;
+
+            buffer = new ReadOnlySpan<byte>(new byte[] { 1, 0, 0, 0 });
+            Assert.True(buffer.TryRead(out value));
+            Assert.Equal(1u, value);
+
+            buffer = new Span<byte>(new byte[] { 1, 0, 0 });
```
Comment (ahsonkhan) [inline tests/System.Slices.Tests/CastTests.cs]: ReadOnlySpan?

File: src/System.Slices/System/SpanExtensions_binary.cs [hunk 5]
```diff
@@ -58,7 +58,37 @@ public static partial class SpanExtensionsLabs
...
+        }
+
+        /// <summary>
+        /// Reads a structure of type T out of a slice of bytes.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TryRead<[Primitive]T>(this ReadOnlySpan<byte> slice, out T value)
+            where T : struct
+        {
+            if (Unsafe.SizeOf<T>() > (uint)slice.Length) {
```
Comment (ahsonkhan) [inline src/System.Slices/System/SpanExtensions_binary.cs]: Should this be unchecked?

```C#
if (Unsafe.SizeOf<T>() > unchecked((uint)slice.Length)) { 
```

Similar to: https://github.com/dotnet/corefx/pull/16416

File: src/System.Slices/System/SpanExtensions_binary.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/System/SpanExtensions_binary.cs]: It does not do anything yet. We have plans to make it into a constraint that would match value types that have no references in them.

File: src/System.Slices/System/SpanExtensions_binary.cs (hunk 3 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/System/SpanExtensions_binary.cs]: Yeah, possibly. Feel free to file an issue to add it

File: src/System.Slices/System/SpanExtensions_binary.cs (hunk 5 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/System/SpanExtensions_binary.cs]: We have a discussion on this. There does not seem to be consensus that we want that. Once we have concensus, there will be thousands of places to change.

File: src/System.Slices/System/SpanExtensions_binary.cs (hunk 3 above)
Comment (ahsonkhan) [inline src/System.Slices/System/SpanExtensions_binary.cs]: Added issue #1253

File: src/System.Slices/System/SpanExtensions_binary.cs (hunk 5 above)
Comment (ahsonkhan) [inline src/System.Slices/System/SpanExtensions_binary.cs]: Sounds good :)

File: src/System.Slices/System/SpanExtensions_binary.cs (hunk 2 above)
Comment (ahsonkhan) [inline src/System.Slices/System/SpanExtensions_binary.cs]: Thanks for clarifying.

File: src/System.Slices/System/SpanExtensions_binary.cs (hunk 1 above)
Comment (jkotas) [inline src/System.Slices/System/SpanExtensions_binary.cs]: I have added issue https://github.com/dotnet/corefx/issues/16479 to `ReadUnaligned<T>`/`WriteUnaligned<T>` APIs on the Unsafe class that will make it possible to write this in portable way.

**Pull request merged**